### PR TITLE
Support multiple export formats and export from entity view

### PR DESCRIPTION
**Pull request by shushu**
#134
#129

Multiple formats on Biblio view:
export3
export2
Multiple formats on entity view:
export1

### CODE REVIEW

File: plugins/biblio_style/citeproc/citeproc.inc [hunk 1]
```diff
@@ -14,5 +14,12 @@ if (module_exists('libraries') && file_exists(libraries_get_path('citeproc-php')
       // Link to the biblio.
       'label_as_link' => TRUE,
     ),
+    'export' => array(
+      'csl' => array(
+        'class' => 'BiblioStyleCIteProc',
```
Comment (amitaibu) [inline plugins/biblio_style/citeproc/citeproc.inc]: Remove.

File: plugins/biblio_style/endnote/endnote.inc [hunk 2]
```diff
@@ -2,8 +2,24 @@
...
   'import' => TRUE,
+  'export' => array(
+    'xml7' => array(
+      'class' => 'BiblioStyleEndNoteXML7',
+      'name' => t('XML7'),
+      'file extension' => 'xml',
+    ),
+    'xml8' => array(
+      'class' => 'BiblioStyleEndNoteXML8',
+      'name' => t('XML8'),
```
Comment (amitaibu) [inline plugins/biblio_style/endnote/endnote.inc]: EndNote - XML

File: plugins/biblio_style/endnote/endnote.inc [hunk 3]
```diff
@@ -2,8 +2,24 @@
 
 $plugin = array(
   'title' => t('EndNote'),
-  'export extension' => 'enl',
   'description' => t('Endnote (tagged or XML) style.'),
   'class' => 'BiblioStyleEndNote',
   'import' => TRUE,
+  'export' => array(
+    'xml7' => array(
```
Comment (amitaibu) [inline plugins/biblio_style/endnote/endnote.inc]: Remove.

File: modules/biblio_ui/biblio_ui.module [hunk 4]
```diff
@@ -1139,21 +1146,31 @@ function biblio_ui_is_biblio_view($view_name, $display_name) {
...
   foreach (biblio_get_exportable_biblio_styles() as $style_name => $style) {
-    $links[$style_name] = array(
-      'href' => 'biblio/export_view/' . $style_name . '/' . $variables['view_name'] . '/' . $variables['display_name'],
-      'title' => $style['title'],
-      'attributes' => array(
-        'title' => t('Click to download Biblio in @format format', array('@format' => $style_name)),
-      ),
-    );
+    foreach ($style['export'] as $format_name => $format) {
+      if (count($style['export']) == 1) {
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: unify `==1` and `>1`

File: biblio.module [hunk 5]
```diff
@@ -227,11 +227,19 @@ function biblio_get_biblio_style($style_name) {
  *
  * @param $style_name
  *   The plugin that should be loaded.
+ * @param null $export_format_name
```
Comment (amitaibu) [inline biblio.module]: remove null

Comment (amitaibu) [inline biblio.module]: Add `(Optional)`

File: biblio.module [hunk 6]
```diff
@@ -227,11 +227,19 @@ function biblio_get_biblio_style($style_name) {
...
+ *   Export format name.
  * @return array
```
Comment (amitaibu) [inline biblio.module:234]: What return. Line break.

File: biblio.module [hunk 7]
```diff
@@ -227,11 +227,19 @@ function biblio_get_biblio_style($style_name) {
...
  */
-function biblio_get_exportable_biblio_style($style_name) {
+function biblio_get_exportable_biblio_style($style_name, $export_format_name = NULL) {
   $style = biblio_get_biblio_style($style_name);
-  if (empty($style['export extension'])) {
+  if (empty($style['export'])) {
+    return;
```
Comment (amitaibu) [inline biblio.module:241]: Explain why return.

File: modules/biblio_ui/biblio_ui.module [hunk 8]
```diff
@@ -67,14 +67,24 @@ function biblio_ui_menu() {
     'weight' => '70',
   );
 
-  $items['biblio/export_view/%/%/%'] = array(
+  $items['biblio/export/%/%/%'] = array(
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: Give example.
`// For example biblio/export/%/%/%`

File: modules/biblio_ui/biblio_ui.module [hunk 9]
```diff
@@ -1106,7 +1123,73 @@ function biblio_ui_export_view($style_name, $view_name, $display_name) {
...
+ * Menu access; Allow export based on existing Biblio style and user access.
+ *
+ * @param $style_name
+ *   Biblio style name.
+ * @param $format_name
+ *   Biblio export format string.
+ * @param $bids_string
+ *   String of comma separated Biblio IDs.
+ *
+ * @return bool
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: Fix.

File: modules/biblio_ui/biblio_ui.module [hunk 10]
```diff
@@ -1139,16 +1222,46 @@ function biblio_ui_is_biblio_view($view_name, $display_name) {
...
+  $attributes = array(
+    'links' => $links,
+    'title' => t('Export'),
+    'attributes' => array('class' => array('biblio-export-buttons')),
+  );
+  return theme('ctools_dropdown', $attributes);
+}
+
+/**
+ * Theme callback - display a button for export in Biblio entity.
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: `Theme callback -` =>  `Theme callback; ...`

File: modules/biblio_ui/biblio_ui.module [hunk 11]
```diff
@@ -1158,3 +1271,23 @@ function theme_biblio_ui_view_export_area($variables) {
   );
   return theme('ctools_dropdown', $attributes);
 }
+
+/**
+ * Implementation of hook_field_extra_fields().
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: `Implementation of` => `Implements`

File: biblio.module [hunk 12]
```diff
@@ -223,15 +224,44 @@ function biblio_get_biblio_style($style_name) {
...
  *
  * @param $style_name
  *   The plugin that should be loaded.
- * @return array
- */
-function biblio_get_exportable_biblio_style($style_name) {
+ * @param $export_format_name
+ *   (Optional) Export format name.
+ *
+ * @return plugin
```
Comment (amitaibu) [inline biblio.module]: Remove plugin

File: biblio.module [hunk 13]
```diff
@@ -223,15 +224,44 @@ function biblio_get_biblio_style($style_name) {
...
+ *   CTools plugin array.
```
Comment (amitaibu) [inline biblio.module]: Biblio style plugin.

File: modules/biblio_ui/biblio_ui.module [hunk 14]
```diff
@@ -66,15 +66,30 @@ function biblio_ui_menu() {
     'type' => MENU_LOCAL_TASK,
     'weight' => '70',
   );
+  // For example: biblio/export/endnote/tagged/1
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: `biblio/export/endnote/1,2,3/tagged`

`biblio/export/bibtex/4,5,6`

File: plugins/biblio_style/bibtex/bibtex.inc [hunk 15]
```diff
@@ -11,4 +10,10 @@ $plugin = array(
     'opening_tag' => '{',
     'closing_tag' => '}',
   ),
+  'export' => array(
+    'bib' => array(
```
Comment (amitaibu) [inline plugins/biblio_style/bibtex/bibtex.inc]: bib => bibtex

File: plugins/biblio_style/citeproc/citeproc.inc [hunk 16]
```diff
@@ -14,5 +14,11 @@ if (module_exists('libraries') && file_exists(libraries_get_path('citeproc-php')
       // Link to the biblio.
       'label_as_link' => TRUE,
     ),
+    'export' => array(
+      'csl' => array(
```
Comment (amitaibu) [inline plugins/biblio_style/citeproc/citeproc.inc]: => citeproc

File: modules/biblio_ui/biblio_ui.module [hunk 17]
```diff
@@ -66,15 +66,30 @@ function biblio_ui_menu() {
...
+    'title' => t('Export Biblio'),
+    'description callback' => 'biblio_ui_description_callback',
+    'description arguments' => array(1, 'export'),
+    'access callback' => 'biblio_ui_export_access',
+    'access arguments' => array(2, 3, 4),
+    'page callback' => 'biblio_ui_export_menu_item',
+    'page arguments' => array(2, 3, 4),
+  );
 
-  $items['biblio/export_view/%/%/%'] = array(
```
Comment (shushu) [inline modules/biblio_ui/biblio_ui.module:70]: biblio/export_view/biblio/page/endnote
biblio/export_view/biblio/page/endnote/tagged
biblio/export_view/endnote/tagged/biblio/page

File: modules/biblio_ui/biblio_ui.module (hunk 14 above)
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]: `biblio/export/1,2,3/endnote/tagged`

`biblio/export/3,4,5/bibtex`

File: modules/biblio_ui/biblio_ui.module (hunk 17 above)
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module:70]: `biblio/export_view/view-name/display-id/style-name/style-format`

File: modules/biblio_ui/biblio_ui.module [hunk 18]
```diff
@@ -1117,7 +1208,8 @@ function biblio_ui_export_view($style_name, $view_name, $display_name) {
  * @param $display_name
  *   View display name.
  *
- * @return bool
+ * @return TRUE if Biblio view and display exist and accessible.
```
Comment (amitaibu) [inline modules/biblio_ui/biblio_ui.module]:
``` php
@return
  TRUE if ...
```

**Comment by amitaibu** [comment]
Merged, thanks.

**Pull request closed without merging**